scanner/decisions/0005-scannerv4-read-cvss-data.md

```diff
@@ -0,0 +1,36 @@
+# 0005 - Adapt Scanner V4 NVD CVSS data updating pipeline to integrate with NVD CVE Api
+
+- **Author(s):** Yi Li <yli3@redhat.com>
+- **Created:** [2023-10-10]
+
+## Status
+
+Accepted
+
+## Context
+
+The National Vulnerability Database (NVD) has announced the retirement of its JSON feeds by December 15, 2023. This creates the need to change our existing CVSS data updater pipeline for Scanner V4, which currently relies on these JSON feeds. 
+
+The NVD JSON feed offers bundled data on a yearly basis, spanning from 2002 to the present year. A benefit is that only two requests are needed in the workflow for each year (one for the meta file and one for the data bundle). However, a limitation is the absence of filtering capabilities, such as by CVSS score or modification date. In contrast, the NVD API allows data retrieval through specific HTTP URL parameters, such as publish date, modified date, or CVSS score, enabling more precise data acquisition.
+
+## Decision
+
+The updated NVD CVSS GitHub Workflow (CI) has transitioned to sourcing data from the NVD CVE API, rather than the NVD Json feeds, ensuring the timely update of the NVD CVSS data bundle in Google Storage. To fetch data, the workflow downloads based on both the start index and CVSS v3 severity. For instance, a command such as curl "https://services.nvd.nist.gov/rest/json/cves/2.0?cvssV3Severity=CRITICAL&startIndex=0" > critical-1.json would be used. For every severity level, the workflow iteratively exhaust the startIndex to capture all associated data. And there are four severity levels in total. After all json files downloaded, they will be compress to one data bundle as one single zip file.
+
+We've chosen to base our API requests on CVSS V3 severity to ensure all data integrated into Claircore possesses valid CVSS V3 metrics. Consequently, retrieving data without CVSS V3 information becomes irrelevant.
+
+The CVSS Updater in Central will be modified to download zip file (generated by CI mentioned above) in the Google bucket. This eliminates the need for JSON parsing in the updater. ScannerV4 will pull the zip bundle from Central and populate the Matcher DB with the enrichment data.
+
+The CVSS Updater in Central will attempt to download the data bundle from the Google bucket a maximum of five times. Should all five attempts be unsuccessful, the updater will pause its efforts and initiate a fresh download attempt after a 4-hour interval. After five unsuccessful download attempts, a failure notification will be swiftly sent to the Scanner on-call Slack channel.
```

Comment:
1. Promising a slack failure notification from Central is a far stretch; Central is deployed in CI, but also in Customer infra, ACSCS, and the wild, and it might be too costly to do that from those environments. Even if possible, you want to know if the bundle endpoint is healthy. We can do that by having something like a Canary (A CI job that pulls the vuln and validates its content). That's a good idea because it allows us to know if customers are experiencing issues when pulling the vuln data themselves.

2. For Central, consider creating a metric for these errors and logs. That might allow us to see those failures and why they happen if customers see the effect (stale vuln data).

3. A better retry strategy might be exponentially backing off with a large max interval (maybe hours). 
   Example:

   1. T=0s
   2. T=10s
   3. T=30s
   4. T=5min
   5. T=15min
   6. T=30min
   7. T=1h
   8. T=1h
   9. T=1h

Saying all that, for the ADR, I think it's better just to lay out these options generally, but don't get too specific unless you're sure you will implement these specifics, which I think it's hard to predict at this point. So just mentioning you are adding metrics, logs, and retrying with large intervals would be good IMO.

Reply:
I agree. "A failure notification will be swiftly sent to the Scanner on-call Slack channel." was too much. I will remove it.

Reply:
@jvdm 
My point is may be we can leave the implementation details of those in the future PR review.

Reply:
@jvdm:

> Saying all that, for the ADR, I think it's better just to lay out these options generally, but don't get too specific unless you're sure you will implement these specifics.

:point_up: 

> I can certainly add that in Central, do we need to specify those in the doc?
> [...]
> For the exponential backoff strategy, I am assuming we don't need to specify in the doc but only to decide in implementation, is that right?

Yes. I added those values to clarify what I meant and to point out that it doesn't make sense to send a notification to Slack from Central when pulling vulns fail.